torchchat/distributed/generate.py

```diff
@@ -0,0 +1,256 @@
+        self,
+        builder_args,
+        generator_args,
+        pipes,
+        loop,
+    ):
+        self.builder_args = builder_args
+        self.generator_args = generator_args
+        self.requests = {}
+        self.in_flight_requests = {}
+        self.in_flight_batch_order = []
+        self.pipes = pipes
+        self.req_to_states = {}
+        self.req_to_results = {}
+        self.request_queue = mp.Queue()
+        self.loop = loop
+
+    def schedule_request(self, req: Request):
+        self.req_to_states[req.request_id] = asyncio.Event()
+        self.req_to_results[req.request_id] = deque()
+        self.request_queue.put(req)
+
+    def process_requests_loop(self):
+        while True:
+            req = self.request_queue.get()
+            if req == "stop":
+                break
+            self.requests = {req.request_id: req.prompt}
+
+            responses = {}
+            running = True
+            while running:
+                outputs = self.step()
+                self.req_to_results[req.request_id].append(outputs[0])
+
+                self.loop.call_soon_threadsafe(self.req_to_states[req.request_id].set)
+
+                running &= not outputs[0].is_finished
+
+    async def wait_for_request(self, req: Request) -> Output:
+        is_finished = False
+        while not is_finished:
+            await self.req_to_states[req.request_id].wait()
+            while len(self.req_to_results[req.request_id]):
+                output = self.req_to_results[req.request_id].popleft()
+                is_finished |= output.is_finished
+                yield output
+        del self.req_to_states[req.request_id]
+        del self.req_to_results[req.request_id]
+
+    def step(self) -> List[Output]:
+        responses = []
+        # TODO: Implement a scheduler to handle the requests
+        if len(self.in_flight_requests) > 0:
+            # Receive decoded token
+            for p in self.pipes:
+                p.send("step")
+            for p in self.pipes:
+                responses.append(p.recv())
+
+        else:
+            # Send requests to backend
+            self.in_flight_batch_order = list(self.requests.keys())
+            prompts = [self.requests[k] for k in self.in_flight_batch_order]
+            for p in self.pipes:
+                p.send(prompts)
+            self.in_flight_requests = self.requests
+            self.requests = {}
+            self.current_step = 0
+            # Receive first token
+            for p in self.pipes:
+                responses.append(p.recv())
+        # Filter out None responses from in-between stages
+        responses = [r for r in responses if r is not None][0]
+        outputs = []
+        for k, v in zip(self.in_flight_batch_order, zip(responses[0], responses[1])):
+            text, token_ids = v
+            outputs.append(
+                Output(
+                    is_finished=self.current_step >= self.generator_args.max_new_tokens,
```

Comment:
maybe TODO here that we need to watch for tokenizer.eos_id token as well as different end of generation?